### PR TITLE
feat: Present Proof - implements the client

### LINKED ISSUE #1560: Present Proof - need to implement the client.
Need to implement the client.

### PR DESCRIPTION
**Pull request by soluchok**
Implements the client for the Present Proof protocol.

Closes #1560 
Signed-off-by: Andrii Soluk <isoluchok@gmail.com>

### CODE REVIEW

File: pkg/client/presentproof/client.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+	errEmptyProposePresentation = errors.New("propose presentation message is empty")
+)
+
+// Provider contains dependencies for the protocol and is typically created by using aries.Context()
+type Provider interface {
+	Service(id string) (interface{}, error)
+}
+
+// ProtocolService defines the presentproof service.
+type ProtocolService interface {
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:27]: nitpick: this doesn't need to be exported

File: pkg/client/presentproof/client.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+		return nil, errors.New("cast service to presentproof service failed")
+	}
+
+	return &Client{
+		Event:   svc,
+		service: svc,
+	}, nil
+}
+
+// Actions returns unfinished actions for the async usage
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:58]:
```suggestion
// Actions returns pending actions that have yet to be executed or cancelled.
```

File: pkg/client/presentproof/client.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+type Provider interface {
+	Service(id string) (interface{}, error)
+}
+
+// ProtocolService defines the presentproof service.
+type ProtocolService interface {
+	service.DIDComm
+	Actions() ([]presentproof.Action, error)
+	ActionContinue(piID string, opt presentproof.Opt) error
+	ActionStop(piID string, err error) error
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:31]: nitpick: I would have placed these two methods on the `Action` objects themselves

File: pkg/client/presentproof/client.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	}, nil
+}
+
+// Actions returns unfinished actions for the async usage
+func (c *Client) Actions() ([]presentproof.Action, error) {
+	return c.service.Actions()
+}
+
+// SendRequestPresentation is used by the Verifier to send a request presentation.
+func (c *Client) SendRequestPresentation(msg *presentproof.RequestPresentation, myDID, theirDID string) error {
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:64]: I have no problems with this, but I notice that this is directly referencing models in the service package, which is a big no-no for reasons unclear to me. Maybe others can better explain why.

File: pkg/client/presentproof/client.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	}
+
+	msg.Type = presentproof.RequestPresentationMsgType
+
+	_, err := c.service.HandleInbound(service.NewDIDCommMsgMap(msg), myDID, theirDID)
+
+	return err
+}
+
+// AcceptRequestPresentation is used when the Prover is willing to accept the request presentation.
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:76]:
```suggestion
// AcceptRequestPresentation is used when the Verifier is willing to accept the request presentation.
```

File: pkg/client/presentproof/client.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+// NOTE: For async usage.
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:77]:
```suggestion

```

File: pkg/client/presentproof/client.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+func (c *Client) AcceptRequestPresentation(piID string, msg *presentproof.Presentation) error {
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:78]: why does the user need to pass back the presentation?

File: pkg/client/presentproof/client.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	return c.service.ActionContinue(piID, WithPresentation(msg))
+}
+
+// NegotiateProposePresentation is used when the Prover wants to negotiate about a request presentation he received.
+// NOTE: For async usage. This function can be used only after receiving a request presentation.
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:83]:
```suggestion

```

File: pkg/client/presentproof/client.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+	return err
+}
+
+// AcceptRequestPresentation is used when the Prover is willing to accept the request presentation.
+// NOTE: For async usage.
+func (c *Client) AcceptRequestPresentation(piID string, msg *presentproof.Presentation) error {
+	return c.service.ActionContinue(piID, WithPresentation(msg))
+}
+
+// NegotiateProposePresentation is used when the Prover wants to negotiate about a request presentation he received.
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:82]:
```suggestion
// NegotiateProposePresentation is used when the Verifier wants to negotiate about a request presentation he received.
```

Comment (llorllale) [inline pkg/client/presentproof/client.go:82]:
```suggestion
// NegotiateProposePresentation is used when the Verifier wants to re-negotiate a presentation proposal they received.
```

File: pkg/client/presentproof/client.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+// NOTE: For async usage. This function can be used only after receiving a request presentation.
+func (c *Client) NegotiateProposePresentation(piID string, msg *presentproof.ProposePresentation) error {
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:84]: Shouldn't this method accept a `RequestPresentation` instead?

File: pkg/client/presentproof/client.go (hunk 9 above)
Comment (llorllale) [inline pkg/client/presentproof/client.go:82]:
```suggestion
// NegotiateProposePresentation is used when the Prover wants to counter a presentation request they received with a proposal.
```

File: pkg/client/presentproof/client.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+	return c.service.ActionContinue(piID, WithProposePresentation(msg))
+}
+
+// DeclineRequestPresentation is used when the Prover does not want to accept the request presentation.
+// NOTE: For async usage.
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:89]:
```suggestion

```

File: pkg/client/presentproof/client.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+	msg.Type = presentproof.ProposePresentationMsgType
+
+	_, err := c.service.HandleInbound(service.NewDIDCommMsgMap(msg), myDID, theirDID)
+
+	return err
+}
+
+// AcceptProposePresentation is used when the Verifier is willing to accept the propose presentation.
+// NOTE: For async usage.
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:108]:
```suggestion

```

File: pkg/client/presentproof/client.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+func (c *Client) AcceptProposePresentation(piID string, msg *presentproof.RequestPresentation) error {
+	return c.service.ActionContinue(piID, WithRequestPresentation(msg))
+}
+
+// DeclineProposePresentation is used when the Verifier does not want to accept the propose presentation.
+// NOTE: For async usage.
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:114]:
```suggestion

```

File: pkg/client/presentproof/client.go [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+func (c *Client) DeclineProposePresentation(piID, reason string) error {
+	return c.service.ActionStop(piID, errors.New(reason))
+}
+
+// AcceptPresentation is used when the Verifier is willing to accept the presentation.
+// NOTE: For async usage.
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:120]:
```suggestion

```

File: pkg/client/presentproof/client.go [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+	return c.service.ActionContinue(piID, WithRequestPresentation(msg))
+}
+
+// DeclineProposePresentation is used when the Verifier does not want to accept the propose presentation.
+// NOTE: For async usage.
+func (c *Client) DeclineProposePresentation(piID, reason string) error {
+	return c.service.ActionStop(piID, errors.New(reason))
+}
+
+// AcceptPresentation is used when the Verifier is willing to accept the presentation.
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:119]:
```suggestion
// AcceptPresentation is used by the Verifier to accept a presentation.
```

File: pkg/client/presentproof/client.go [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+// NOTE: For async usage.
+func (c *Client) AcceptPresentation(piID string) error {
+	return c.service.ActionContinue(piID, nil)
+}
+
+// DeclinePresentation is used when the Verifier does not want to accept the presentation.
+// NOTE: For async usage.
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:126]:
```suggestion

```

File: pkg/client/presentproof/client.go [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+
+// ProtocolService defines the presentproof service.
+type ProtocolService interface {
+	service.DIDComm
+	Actions() ([]presentproof.Action, error)
+	ActionContinue(piID string, opt presentproof.Opt) error
+	ActionStop(piID string, err error) error
+}
+
+// Client enable access to presentproof API
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:34]: Please include a link to the RFC

File: pkg/client/presentproof/client.go [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+// DeclinePresentation is used when the Verifier does not want to accept the presentation.
+// NOTE: For async usage.
+func (c *Client) DeclinePresentation(piID, reason string) error {
+	return c.service.ActionStop(piID, errors.New(reason))
+}
+
+// WithPresentation allows providing Presentation message
+// USAGE: This message can be provided after receiving a Request message
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:132]:
```suggestion
// Use this option to respond to RequestPresentation
```

File: pkg/client/presentproof/client.go [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+func WithPresentation(msg *presentproof.Presentation) presentproof.Opt {
+	return presentproof.WithPresentation(msg)
+}
+
+// WithProposePresentation allows providing ProposePresentation message
+// USAGE: This message can be provided after receiving a Request message
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:138]:
```suggestion
// Use this option to respond to RequestPresentation
```

Comment (llorllale) [inline pkg/client/presentproof/client.go:138]:
```suggestion
// Use this option to respond to RequestPresentation
```

File: pkg/client/presentproof/client.go [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+func WithProposePresentation(msg *presentproof.ProposePresentation) presentproof.Opt {
+	return presentproof.WithProposePresentation(msg)
+}
+
+// WithRequestPresentation allows providing RequestPresentation message
+// USAGE: This message can be provided after receiving a propose message
```
Comment (llorllale) [inline pkg/client/presentproof/client.go:144]:
```suggestion
// Use this option to respond to ProposePresentation
```

File: pkg/client/presentproof/client.go (hunk 1 above)
Comment (soluchok) [reply pkg/client/presentproof/client.go:27]: I am using it for mock generation e.g the usage `mocks.NewMockProtocolService(ctrl)`

File: pkg/client/presentproof/client.go (hunk 4 above)
Comment (soluchok) [reply pkg/client/presentproof/client.go:64]: Here is my answer. https://github.com/hyperledger/aries-framework-go/pull/1480#discussion_r395142761

Doing something like this is not a solution.
```
type RequestPresentation struct {
	*presentproof.RequestPresentation
}

```

Comment (soluchok) [reply pkg/client/presentproof/client.go:64]: Ok, actually there is one way.

```
type RequestPresentation presentproof.RequestPresentation
// and then we can easily convert it
r1 := presentproof.RequestPresentation(*r)
fmt.Println(&r1)
```

Comment (soluchok) [reply pkg/client/presentproof/client.go]: @llorllale  Done!

File: pkg/client/presentproof/client.go (hunk 10 above)
Comment (soluchok) [reply pkg/client/presentproof/client.go]: Done! `NegotiateRequestPresentation` sounds good to me.

File: pkg/client/presentproof/client.go (hunk 3 above)
Comment (soluchok) [reply pkg/client/presentproof/client.go:42]: These functions are supposed to be used for async usage (rest API).  We need to continue with the protocol by `piID` (not by the action object). `ActionContinue` and `ActionStop` are using `service` under the hood. If I add these methods it means we will keep `service` in memory. It makes things complicated.

File: pkg/client/presentproof/client.go (hunk 4 above)
Comment (llorllale) [reply pkg/client/presentproof/client.go]: @soluchok that does not address the point @rolsonquadras was (unsuccessfully) attempting to make in https://github.com/hyperledger/aries-framework-go/pull/1480#discussion_r395124308. These are only half-measures: true decoupling of the client and service packages would require more.

But like I said in my original comment above: I didn't have a problem with your original design anyway.

**Pull request merged**